### PR TITLE
Reader: Teach the feed reducer and selector how to fetch stale feeds

### PR DESCRIPTION
**Pull request by blowery**
This teaches the feed reducer / selector / query component how to refetch stale feed data. Important points:

* Feed data is stale if it's more than a day old, or if we don't know when we last fetched it. 
* We're not serializing the update date, so a page refresh will also refresh all of the feed data. We'll have feed data for the initial render, but updated data will come in in the background.
* This also makes the shape of the feed data more explicit, removing the `meta` property which we don't actually use anywhere. This saves a bit of space on disk.

To test, load up the reader, then following manage. Then reload. Should see feed names immediately, but you should also see network requests in the background updating the data.

### CODE REVIEW

File: client/state/reader/feeds/reducer.js [hunk 1]
```diff
@@ -54,28 +56,26 @@ function handleRequestFailure( state, action ) {
...
 	} );
 }
 
 function handleFeedUpdate( state, action ) {
-	const feeds = map( action.payload, feed => {
-		feed = assign( {}, feed );
-		adaptFeed( feed );
-		return feed;
-	} );
+	const feeds = map( action.payload, feed => adaptFeed( feed ) );
```
Comment (samouri) [inline client/state/reader/feeds/reducer.js]: `map( action.payload, adaptFeed );`

File: client/state/reader/feeds/schema.js [hunk 2]
```diff
@@ -12,8 +12,7 @@ export const itemsSchema = {
 				URL: { type: [ 'string', 'null' ] },
 				feed_URL: { type: [ 'string', 'null' ] },
 				is_following: { type: [ 'boolean', 'null' ] },
-				subscribers_count: { type: [ 'integer', 'null' ] },
-				meta: { type: [ 'object', 'null' ] }
```
Comment (samouri) [inline client/state/reader/feeds/schema.js]: what was meta for?

Comment (blowery) [inline client/state/reader/feeds/schema.js]: just completeness. we were not eliding it before.

**Comment by samouri** [comment]
this is a pre-req for: https://github.com/Automattic/wp-calypso/pull/13836

**Comment by blowery** [comment]
> this is a pre-req in my mind for: #13836

this _is_ #13836 :D

**Comment by blowery** [comment]
If we like how this looks, I'll dupe it for sites.

**Comment by samouri** [comment]
can we rename lastUpdated to something like lastFetched? I think lastUpdates is going to get overloaded because of the new ui piece involving it

File: client/state/reader/feeds/selectors.js [hunk 3]
```diff
@@ -1,14 +1,29 @@
...
 *
 * @param  {Object}  state  Global state tree
 * @param  {Number}  feedId The feed ID
 * @return {Boolean}        Whether feed should be fetched
 */
-
 export function shouldFeedBeFetched( state, feedId ) {
-	return ! state.reader.feeds.queuedRequests[ feedId ] && // not currently queued
-		! state.reader.feeds.items[ feedId ]; // not currently loaded
+	// we should fetch the feed if we don't have it,
```
Comment (samouri) [inline client/state/reader/feeds/selectors.js]: you can drop the comments if you make obvious vars for it all:
```js
const alreadyRequesting = ! state.reader.feeds.queuedRequests[ feedId ];
const staleFeed = ! isStale( state, feedId, DAY_IN_MILLIS );
const missingFeed = ! getFeed( state, feedId );

return ( ! alreadyRequesting ) && ( staleFeed || missingFeed )
```

File: client/state/reader/feeds/selectors.js [hunk 4]
```diff
@@ -1,14 +1,29 @@
...
+	// or we do have it
+	// and it's been more than a day since we last fetched, and it's not already queued to be fetched
+	return ! state.reader.feeds.queuedRequests[ feedId ] &&
+		(
+			! getFeed( state, feedId ) ||
+			! lastUpdatedWithin( state, feedId, DAY_IN_MILLIS )
+		);
+}
+
+function lastUpdatedWithin( state, feedId, timeInMillis ) {
```
Comment (samouri) [inline client/state/reader/feeds/selectors.js]: FFTI:
maybe make it return the flip of this with either something like isStale, isStalerThan, isOlderThan?

**Comment by samouri** [comment]
logic wise it all looks good.  Also really smart. 
theres one higher level thought I was having that we don't need to iron out in this PR:
-  there is a lot of duplication between sites+ feeds redux logic.  how could we remove the duplication while maintaining extensibility?

**Comment by samouri** [comment]
tested:
- [ ] clicked around a bunch. things worked as expected.
- [ ] saw that feeds were not refreshed when i click between manage and site-stream
- [ ] manually set the lastUpdated to 0, then went to site-stream.  was refreshed as expected.

File: client/state/reader/feeds/selectors.js (hunk 4 above)
Comment (blowery) [inline client/state/reader/feeds/selectors.js]: yeah, all the negation was bugging me too. I'll flip it.

File: client/state/reader/feeds/selectors.js (hunk 3 above)
Comment (blowery) [inline client/state/reader/feeds/selectors.js]: nice, i like it

**Comment by blowery** [comment]
> can we rename lastUpdated to something like lastFetched? I think lastUpdates is going to get overloaded because of the new ui piece involving it

I originally had it as `lastFetched` but then added in the bit that updates time when we get `*_UPDATE` actions, which are not exactly "fetches". Though I guess it is a fetch, it's just done via meta... But I can flip it back. Not a biggie.

**Comment by blowery** [comment]
> there is a lot of duplication between sites+ feeds redux logic. how could we remove the duplication while maintaining extensibility?

Yeah, the trick there is that if we drop the duplication, I'd want to do it in a sort of second-order way. First order would just be extracting the exact duplication between feeds and sites. That's nice but doesn't have a ton of utility. Second order would be "what's common between this and other entity fetches that may need staleness". Which to me sounds more like `needs`.

File: client/state/reader/feeds/selectors.js [hunk 5]
```diff
@@ -7,32 +7,29 @@ const DAY_IN_MILLIS = 24 * 60 * 1000 * 1000;
...
-	// or we do have it
-	// and it's been more than a day since we last fetched, and it's not already queued to be fetched
-	return ! state.reader.feeds.queuedRequests[ feedId ] &&
-		(
-			! getFeed( state, feedId ) ||
-			! lastUpdatedWithin( state, feedId, DAY_IN_MILLIS )
-		);
+	const isNotQueued = ! state.reader.feeds.queuedRequests[ feedId ];
+	const isMissing = ! getFeed( state, feedId );
+	return isNotQueued && ( isMissing || isStale( state, feedId ) );
```
Comment (samouri) [inline client/state/reader/feeds/selectors.js]: love it

File: client/state/reader/feeds/selectors.js [hunk 6]
```diff
@@ -7,17 +9,27 @@
...
 }
 
 /**
- * Get the feed object for the given feed id
- *
- * @param  {Object} state  Global state tree
- * @param  {Number} feedId The feed ID
- * @return {Object}        The feed object
- */
+* Returns a feed object
```
Comment (samouri) [inline client/state/reader/feeds/selectors.js]: spacing here got garbled i think

**Comment by samouri** [comment]
give it another round of testing!

**Pull request merged**